src/Microsoft.DotNet.DeadRegionAnalysis/AnalysisEngine.RegionRemoval.cs

```diff
@@ -0,0 +1,196 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Linq;
+using System.Text;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.CSharp;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using Microsoft.CodeAnalysis.Text;
+
+namespace Microsoft.DotNet.DeadRegionAnalysis
+{
+    public partial class AnalysisEngine
+    {
+        public async Task<Document> RemoveUnnecessaryRegions(DocumentConditionalRegionInfo info, CancellationToken cancellationToken = default(CancellationToken))
+        {
+            if (info == null)
+            {
+                throw new ArgumentNullException("info");
+            }
+
+            var changes = CalculateTextChanges(info.Chains);
+            if (changes == null || changes.Count == 0)
+            {
+                return info.Document;
+            }
+
+            var newText = await info.Document.GetTextAsync(cancellationToken);
+
+            try
+            {
+                newText = newText.WithChanges(changes);
+            }
+            catch (Exception)
+            {
+                var changesString = new StringBuilder();
+                var syntaxTree = info.Document.GetSyntaxTreeAsync(cancellationToken).Result;
+
+                foreach (var change in changes)
+                {
+                    var lineSpan = Location.Create(syntaxTree, change.Span).GetLineSpan();
+                    changesString.AppendFormat("({0}-{1}): {2}", lineSpan.StartLinePosition.Line, lineSpan.EndLinePosition.Line, newText.GetSubText(change.Span).ToString());
+                }
+
+                Console.WriteLine(string.Format("Failed to remove regions from document '{0}':{1}{2}", info.Document.FilePath, Environment.NewLine, changesString.ToString()));
```

Comment:
If this is meant to be re-hostable we should avoid the console here and instead abstract it out to an interface.